src/Validator/HttpUserAgent.php

```diff
@@ -5,27 +5,18 @@
 namespace Laminas\Session\Validator;
 
 /**
- * @implements ValidatorInterface<string>
+ * @implements ValidatorInterface<string|null>
  */
-class HttpUserAgent implements ValidatorInterface
+final class HttpUserAgent implements ValidatorInterface
 {
-    /**
-     * Internal data
-     *
-     * @var string
-     */
-    protected $data;
-
     /**
      * Constructor
      * get the current user agent and store it in the session as 'valid data'
-     *
-     * @param string|null $data
      */
-    public function __construct($data = null)
+    public function __construct(protected EnvironmentValueObject $env, protected ?string $data = null)
```

Comment:
```suggestion
    public function __construct(private readonly EnvironmentValueObject $initial)
```

The environment passed to the constructor should represent the values you wish to compare against, so, they will probably be retrieved from the session - i.e. it's the data set that was recorded when the session was first started.

I don't know how these validators are supposed to work - if they are constructed also with the to-be-validated data, then the constructor signature would be:

```php
public function __construct(
    private readonly Environment $initial,
    private readonly Environment $current,
)
```

_(Perhaps)_

Then `isValid` is simply `return $this->initial->userAgent === $this->current->userAgent`

Reply:
Also, why do the validators have a `getData` method? - if this is only used in tests, it should be removed, it has no purpose.